pkg/queue/breaker_test.go

```diff
@@ -79,142 +65,140 @@ func TestBreakerInvalidConstructor(t *testing.T) {
-	b := NewBreaker(params)                // Breaker capacity = 2
-	want := []bool{true, true, true, true} // Only two requests will be in flight at a time
-	locks := make([]request, 4)
-	locks[0] = b.concurrentRequest(context.Background())
-	locks[1] = b.concurrentRequest(context.Background())
-	unlock(locks[0])
-	locks[2] = b.concurrentRequest(context.Background())
-	unlock(locks[1])
-	locks[3] = b.concurrentRequest(context.Background())
-	unlockAll(locks[2:])
-
-	if diff := cmp.Diff(accepted(locks), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
-	}
+	// Process the remainder successfully.
+	reqs.processSuccessfully(t)
+	reqs.processSuccessfully(t)
 }
 
 func TestBreakerRecover(t *testing.T) {
 	params := BreakerParams{QueueDepth: 1, MaxConcurrency: 1, InitialCapacity: 1}
-	b := NewBreaker(params)                              // Breaker capacity = 2
-	want := []bool{true, true, false, false, true, true} // Shedding will stop when capacity opens up
-
-	locks := b.concurrentRequests(context.Background(), 4)
-	unlockAll(locks)
-	// Breaker recovers
-	moreLocks := b.concurrentRequests(context.Background(), 2)
-	unlockAll(moreLocks)
-
-	if diff := cmp.Diff(accepted(append(locks, moreLocks...)), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
-	}
+	b := NewBreaker(params) // Breaker capacity = 2
+	reqs := newRequestor(b)
+
+	// Overshoot breaker capacity by 2.
+	reqs.request()
+	reqs.request()
+	reqs.request()
+	reqs.request()
+
+	// Expect 2 failures and 2 successes.
+	reqs.expectFailure(t)
+	reqs.expectFailure(t)
+	reqs.processSuccessfully(t)
+	reqs.processSuccessfully(t)
+
+	// As capacity has freed, more requests should succeed.
+	reqs.request()
+	reqs.request()
+	reqs.processSuccessfully(t)
+	reqs.processSuccessfully(t)
 }
 
 func TestBreakerLargeCapacityRecover(t *testing.T) {
 	params := BreakerParams{QueueDepth: 5, MaxConcurrency: 45, InitialCapacity: 45}
-	b := NewBreaker(params)   // Breaker capacity = 50
-	want := make([]bool, 150) // Process 150 requests
-	for i := 0; i < 50; i++ {
-		want[i] = true // First 50 will fill the breaker capacity.
-	}
-	for i := 50; i < 100; i++ {
-		want[i] = false // The next 50 will be shed.
-	}
-	for i := 100; i < 150; i++ {
-		want[i] = true // The next 50 will be processed as capacity opens up.
-	}
-
-	// Send 100 requests.
-	locks := b.concurrentRequests(context.Background(), 100)
-	// Process one request and send one request, 50 times.
-	for i := 100; i < 150; i++ {
-		// Open capacity
-		unlock(locks[i-100])
-		// Add another request
-		locks = append(locks, b.concurrentRequest(context.Background()))
-	}
-	unlockAll(locks[50:])
+	b := NewBreaker(params) // Breaker capacity = 50
+	reqs := newRequestor(b)
 
-	if diff := cmp.Diff(accepted(locks), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
+	// Overshoot the capacity twice.
+	for i := 0; i < 100; i++ {
```

Comment:
you can parametrize the loops based on `params.qd + params.ic`.
But bigger questions is do we even need this test? Or do we need previous.

Reply:
I agree, ditched.